toy_2d/src/ilqr_2d.py

```diff
@@ -0,0 +1,418 @@
+        """
+        return 0.5*(xf - self.x_goal).T @ self.Qf @ (xf - self.x_goal)
+
+    def grad_terminal_cost(self, xf: np.ndarray) -> np.ndarray:
+        """
+        :param xf: final state
+        :return: ∂Lf/∂xf
+        """
+
+        grad = np.zeros((self.nx))
+
+        # TODO: Compute the gradient
+        x_grad = (xf- self.x_goal).T @ self.Qf
+        grad = x_grad
+        return grad
+        
+    def hess_terminal_cost(self, xf: np.ndarray) -> np.ndarray:
+        """
+        :param xf: final state
+        :return: ∂²Lf/∂xf²
+        """ 
+
+        H = np.zeros((self.nx, self.nx))
+        H = self.Qf
+
+        # TODO: Compute H
+
+        return H
+
+    def forward_pass(self, xx, uu, dd, KK, lr):
+        """
+        :param xx: list of states, should be length N
+        :param uu: list of inputs, should be length N-1
+        :param dd: list of "feed-forward" components of iLQR update, should be length N-1
+        :param KK: list of "Feedback" LQR gain components of iLQR update, should be length N-1
+        :return: A tuple (xtraj, utraj) containing the updated state and input
+                 trajectories after applying the iLQR forward pass
+        """
+
+        xtraj = [np.zeros((self.nx,))] * self.N
+        utraj = [np.zeros((self.nu,))] * (self.N - 1)
+        xtraj[0] = xx[0]
+
+        # TODO: compute forward pass
+        for i in range(self.N-1):
+            utraj[i] = uu[i] + KK[i]@(xtraj[i] - xx[i]) + lr*dd[i]
+            # if(utraj[i][0]<0 and utraj[i][1]>0): utraj[i] = np.array([0,0])
+            #TO DO
+            # xtraj[i+1] = quad_sim.F(xtraj[i], utraj[i], self.dt)
+            xtraj[i+1] = self.sim_forward(xtraj[i], utraj[i])
+        return xtraj, utraj
+
+    def get_linearized_discrete_dynamics(self, state, action):
+        q = np.array([state[0], state[2], state[4]])
+        v = np.array([state[1], state[3], state[5]])
+        lcs.set_linearization_point(q,v, action)
+        A, B_gen = lcs.get_A_matrix(), lcs.get_B_matrix()
+        pmat = lcs.get_P_matrix()
+        B = B_gen@pmat
+        C_mat = lcs.get_C_matrix()
+        G_mat = lcs.get_G_matrix()
+        J_mat = lcs.get_J_matrix()
+        H_mat = lcs.get_H_matrix()
+        #access original dynamics to get the original lambdas
+        curr_lam = self.get_lamda(state, action)
+        eps = 1e-3
+        # check if there is a contact at all. if not use the normal dynamics itself as cube is in air
+        if(np.argwhere(curr_lam>eps).shape[0]>0):
+            curr_lam = curr_lam[curr_lam>eps]
+            # print(curr_lam)
+            C_mat = C_mat[:, np.where(curr_lam>eps)[0]]
+            J_mat = J_mat[:, np.where(curr_lam>eps)[0]]
+            J_pinv = np.linalg.pinv(J_mat)
+            A_d, B_d = -C_mat@J_pinv@G_mat, -C_mat@J_pinv@H_mat@pmat
+            A, B = A + A_d, B+B_d
+
+        return A,B
+    
+    def get_lamda(sels, state, action):
+        system.set_initial_state(state)
```

Comment:
This also just references an externally-defined `system` object, though it should be an attribute of the `iLQR` object instead.